argoproj-labs/argocd-operator
# feat: Reconcile on change of argocd-repo-server-tls and implement AutoTLS feature

**What type of PR is this?**

> /kind enhancement

**What does this PR do / why we need it**:

This PR introduces two new configurables:

* `.spec.repo.stricttls: <bool>` : If set to `true`, will configure `<instance>-server` and `<instance>-application-controller` to validate the TLS certificate of the `<instance>-repo-server`'s gRPC endpoint, by adding `--repo-server-strict-tls` to the pods command's arguments.
* `.spec.repo.autotls: <string>`: Can be set to a provider which does automatic provisioning of the `argocd-repo-server-tls` secret. Currently, the only provider supported is `openshift`

The main feature supported is to watch for changes to the `argocd-repo-server-tls` secret, containing the TLS configuration (cert/key combination) for the gRPC API of the repository server. If a change is detected, the appropriate workloads will be restarted. If the `argocd-repo-server-tls` was created using `AutoTLS` feature, no further configuration will be necessary. If the secret was manually added to the namespace, it will require an annotation `argocds.argoproj.io/name: <instance_name>` to be successfully mapped back to the managing `ArgoCD` instance.

Few words about `AutoTLS` feature on OpenShift:

OpenShift Container Platform provides a cluster-internal CA to issue service certificates, as described [here](https://docs.openshift.com/container-platform/4.7/security/certificates/service-serving-certificate.html#understanding-service-serving_service-serving-certificate). If `.spec.repo.autotls: openshift` is specified, the controller will add the appropriate annotation to the `<instance>-repo-server` Service in the instance's namespace, which indicates that a TLS certificate for this service should be issued by the service CA.

**Have you updated the necessary documentation?**


**Which issue(s) this PR fixes**:

Fixes #?

**How to test changes / Special notes to the reviewer**:

To test the `stricttls` feature (can be tested on any platform, plain K8s and OpenShift alike)

* Create ArgoCD CR with all default values:
  ```yaml
   apiVersion: argoproj.io/v1alpha1
   kind: ArgoCD
   metadata:
     name: argocd
  ```
* Wait until reconciliation is complete and workloads became available

* Modify the ArgoCD CR to set `.spec.repo.stricttls` to `true`:
  ```
  kubectl patch argocds.argoproj.io argocd --type=merge -p '{"spec":{"repo":{"stricttls":true}}}'
  ```
* Check `argocd-server` deployment for containing `--repo-server-strict-tls` in the container's command:
  ```
  kubectl get deployments.apps argocd-server -o jsonpath='{.spec.template.spec.containers[0].command}'
  ```
* Check `argocd-application-controller` statefulset for containing `--repo-server-strict-tls` in the container's command:
  ```
  kubectl get statefulsets.apps argocd-application-controller -o jsonpath='{.spec.template.spec.containers[0].command}'
  ```
* Pods for `argocd-server` and `argocd-application-controller` should have been restarted by that change:
  ```
  kubectl get pods
  ```

To test the `autotls` feature (works only on OpenShift):

* Create ArgoCD CR with all default values:
  ```yaml
   apiVersion: argoproj.io/v1alpha1
   kind: ArgoCD
   metadata:
     name: argocd
  ```
* Wait until reconciliation is complete and workloads became available
* Modify the ArgoCD CR to set `.spec.repo.autotls` to `openshift`:
  ```
  kubectl patch argocds.argoproj.io argocd --type=merge -p '{"spec":{"repo":{"autotls":"openshift"}}}'
  ```
* Annotate the `argocd-repo-server` service so that the OpenShift service CA will issue a certificate into `argocd-repo-server-tls`:
  ```
  kubectl annotate service argocd-repo-server service.beta.openshift.io/serving-cert-secret-name=argocd-repo-server-tls
  ```
* Verify the `argocd-repo-server-tls` secret has been created by the CA's controller:
  ```
  kubectl get secrets argocd-repo-server-tls
  ```
  The secret should be of type `kubernetes.io/tls` and should have a size of `2`
* Verify the workloads `argocd-server`, `argocd-repo-server` and `argocd-application-controller` have been restarted
* For each pod of `argocd-server`, `argocd-repo-server` and `argocd-application-controller`, do a `kubectl exec -it <podname> bash` and look for the following files to exist there:
  * argocd-server: `/app/config/server/tls/tls.crt`
  * argocd-repo-server: `/app/config/reposerver/tls/tls.crt`
  * argocd-application-controller: `/app/config/controller/tls/tls.crt`

To test reconciliation of `argocd-repo-server-tls` on non-OpenShift platform, or without `autotls` enabled:

* Create ArgoCD CR with all default values:
  ```yaml
   apiVersion: argoproj.io/v1alpha1
   kind: ArgoCD
   metadata:
     name: argocd
  ```
* Wait until reconciliation is complete and workloads became available
* Create a self-signed certificate for the repository server (crt and key will be stored in `/tmp/reposerver.crt` and `/tmp/reposerver.key` respectively):
  ```
  openssl req -new -x509 -sha256 \
    -subj "/C=XX/ST=XX/O=Testing/CN=repo-server" \
    -reqexts SAN -extensions SAN \
    -config <(printf "\n[SAN]\nsubjectAltName=DNS:argocd-repo-server.$NAMESPACE.svc.cluster.local\n[req]\ndistinguished_name=req") \
    -keyout /tmp/reposerver.key \
    -out /tmp/reposerver.crt \
    -newkey rsa:4096 \
    -nodes \
    -sha256 \
    -days 10
  ```
* Create the `argocd-repo-server-tls` secret of type `kubernetes.io/tls`:
  ```
  kubectl create secret tls argocd-repo-server-tls --key=/tmp/reposerver.key --cert=/tmp/reposerver.crt
  ```
* Check that pods of `argocd-server`, `argocd-repo-server` and `argocd-application-controller` were *not* restarted just yet
* Annotate the `argocd-repo-server-tls` secret to indicate it belongs to the `argocd` operand:
  ```
  kubectl annotate secret argocd-repo-server-tls argocds.argoproj.io=argocd
  ```
* Check that the Operator reconciled and the pods of `argocd-server`, `argocd-repo-server` and `argocd-application-controller` have been restarted.
